### ISSUE TITLE
Provider feedback

### ISSUE THREAD

**Issue by dmitri-d** [kayak/smart_proxy_dns_menandmice#1]
The provider looks very good, I only have a few notes, mostly regarding unit tests.

- I don't see that "mock_login" parameter of ```Proxy::Dns::Menandmice::Record``` is being used anywhere. I'd remove it, esp. considering it's better not to mix production and testing-related code. If you do need this functionality for testing, I'd suggest creating a test-specific subclass of ```Proxy::Dns::Menandmice::Record```.

- I'd move client initialization out of  ```Proxy::Dns::Menandmice::Record``` constructor, and inject it instead. Personally, I dislike when a call to a constructor is not guaranteed to succeed. More importantly, this would make testing easier. Such a change would allow you to inject a test double in the constructor in your tests. This would allow you to mock/stub calls at the client api level, as opposed to http transport, which would be both more strict and less fragile.

Please let me know if you have any questions.

**Comment by NeilHanlon** [comment]
@witlessbird Where would you recommend pulling client initialization into? I thought of putting it `dns_mendandmice_configuration.rb` in the `load_dependency_injection_wirings`, which is I believe where you suggested, but I'm confused as how to perform tests with this, as it will always try to run `client.login` -- which I suppose I could stub? but I'm not sure that's best

**Comment by NeilHanlon** [comment]
ie,

```ruby
module ::Proxy::Dns::Menandmice
  class PluginConfiguration
    def load_classes
      require 'dns_common/dns_common'
      require 'smart_proxy_dns_menandmice/dns_menandmice_main'
      include Proxy::Log
    end

    def load_dependency_injection_wirings(container_instance, settings)

      begin
        client = MmJsonClient::Client.new(server: settings[:server],
                                         username: settings[:username],
                                         password: settings[:password],
                                         ssl: settings[:ssl],
                                         verify_ssl: settings[:verify_ssl]
        )
        client.login
        container_instance.dependency :dns_provider, (lambda do
          ::Proxy::Dns::Menandmice::Record.new(
              client,
              settings[:dns_ttl])
        end)
      rescue MmJsonClient::ServerConnectionError, MmJsonClient::ServerError => e
        logger.fatal("Could not login to Men & Mice server", e)
        raise e
      end

    end
  end
end
```

**Comment by dmitri-d** [comment]
It depends on whether you want to reuse the client connection between requests. If you do, I would suggest creating a small helper that would return an initialized and logged-in client. Something like:

```
class MmClientInitializer
  extend ::Proxy::Log
  def self.initialized_client(server, username, password, ...)
    client = MmJsonClient::Client.new(server: settings[:server],
                                         username: settings[:username],
                                         password: settings[:password],
                                         ssl: settings[:ssl],
                                         verify_ssl: settings[:verify_ssl])
    client.login
    client
  rescue MmJsonClient::ServerConnectionError, MmJsonClient::ServerError => e
    logger.fatal("Could not login to Men & Mice server", e)
    raise e
  end
end
```

then you'd wire it up as:

```
  container_instance.singleton_dependency :mm_client, lambda { MmClientInitializer.initialized_client }
  container_instance.dependency :dns_provider, (lambda do
    ::Proxy::Dns::Menandmice::Record.new(container_instance.get_dependency(:mm_client, ...)
  end)
```

Otherwise create an instance of MmJsonClient::Client (without logging it in), inject it into ::Proxy::Dns::Menandmice::Record, and call login on it before making any api calls.

Does this make sense?

**Comment by NeilHanlon** [comment]
Yeah, that makes a lot of sense. Thanks!

**Comment by NeilHanlon** [comment]
Can you check out 136ce983c05d734a211208c095e9982b68757c41 and 96c9b25a9a8b5ceb52aa3327de6572546ff9379c ? I think these improvements are what you're looking for.

(In retrospect I probably should've commited these as a branch and merged them in.. oops)

**Comment by dmitri-d** [comment]
Nods, that looks good.

In retrospect I should've mentioned that you'd need to check for potential concurrency issues if you decide to reuse the client instance. Probably worth checking the docs/forums and/or try to spawn a bunch of threads and hit the same api end-point(s) simultaneously...

A few more notes:

- I don't think you need this line: https://github.com/kayak/smart_proxy_dns_menandmice/blob/96c9b25a9a8b5ceb52aa3327de6572546ff9379c/lib/smart_proxy_dns_menandmice/dns_menandmice_configuration.rb#L8

- In test setup (https://github.com/kayak/smart_proxy_dns_menandmice/blob/96c9b25a9a8b5ceb52aa3327de6572546ff9379c/test/dns_menandmice_record_test.rb#L10) you don't actually need to instantiate a real client object anymore. Instead, ```@client = mock()``` or even ```@client = Object.new``` will work.

Then in your tests you can do something like:

```
def test_create_a
   @client.stubs(:match_zone).returns(MmJsonClient::DNSZone.new(...))
    @client.expects(:add_dns_record).with(MmJsonClient::DNSRecord.new(name: ...)).returns("")
    @provider.do_create('test.example.com', '10.1.1.1', 'A')
end
```
The code above assumes that #match_zone has dedicated tests and I'm haven't checked what the actual class for dns zones is.

Factored this way, the test exercises #do_create (atm a mock is setup for the method under test, and then the method is called, which will always pass). Another benefit is that you can verify how method works in normal conditions and under failure. I think a good rule of thumb for mocking is to mock calls as close as possible to code being verified. While httpmock is useful, it's more applicable for higher-level (think integration) tests.

Let me know if you have any questions!

**Comment by NeilHanlon** [comment]
Ah, I think I'm beginning to understand a bit more. I'll play around with that a bit.

**Comment by NeilHanlon** [comment]
Thanks again for the help and guidance. Does this look better?

660d7d4b30a72742d5a385f5e33b62a834eaf300

**Comment by dmitri-d** [comment]
I would suggest taking a look at Martin Fowler's "Mocks aren't Stubs" post (https://martinfowler.com/articles/mocksArentStubs.html). It's a long read, but it gives an exhaustive comparison of various approaches to state vs. behaviour testing and where mocks and stubs fit in.

TLDR version:  stubs are enablers of behaviour testing, but to actually test some behaviour one needs to use mocks -- test doubles that specify and verify expectations. At the moment ```DnsMenandmiceRecordTest``` stubs all possible calls a test can make, leading to suit of tests where every test will always pass. Something like this would verify that expected behaviour happened:

```
def setup
    @client = mock('MmJsonClient::Client')
    @ttl = 999
    @provider = ::Proxy::Dns::Menandmice::Record.new(@client,@ttl)
    @zone = stub('a_zone', :ref => '567')
end

def test_create_a
    expected_name = 'test.example.com'
    expected_address = '10.1.1.1'
    expected_type = 'A'
    record = MmJsonClient::DNSRecord.new(:name => expected_name, :type = expected_type, :ttl => @ttl, :data => expected_address, :enabled => true, :dns_zone_ref => @zone.ref)
    response = stub('a_response', :ref => '567')

    @provider.stubs(:match_zone).returns(@zone)
    @client.expects(:add_dns_record).with(has_entry(:dns_record, record)).returns(response)
    assert_equal response, @provider.do_create('test.example.com', '10.1.1.1', 'A')
end
```

I would suggest playing with different expectations to see how and when the test above fails and passes.

I hope this helps!

**Comment by NeilHanlon** [comment]
Getting some time to look at this again. I appreciate the link--that certainly helped my understanding a whole bunch.

One hurdle I'm running into is when mocking the `add_dns_record` function with an already-instantiated `MmJsonClient::DNSRecord` class, the mock doesn't match the record being created when running the test. I get something like this (Note the different memory object IDs for the class instance):

```
unexpected invocation: #<Mock:MmJsonClient::Client>.add_dns_record(:dns_record => #<MmJsonClient::DNSRecord:0x40eb4c0>)
unsatisfied expectations:
- expected exactly once, not yet invoked: #<Mock:MmJsonClient::Client>.add_dns_record(has_entry(:dns_record => #<MmJsonClient::DNSRecord:0x2de04e8>))
```

I got around this by using the following, ie, checking if `add_dns_record` is called with an instance of `MmJsonClient::DNSRecord`, instead of a specific instance. Is that still a valid test? It would appear to me to be since the instantiated `record` is never used outside of that mock.

```ruby
  def setup
    @client = mock('MmJsonClient::Client')
    @ttl = 999
    @provider = ::Proxy::Dns::Menandmice::Record.new(@client, @ttl)

    @zone = stub('a_zone', :ref => '567', :type => 'Master', :name => 'example.com')
    @zones = stub('a_zonelist', :dns_zones => [@zone])
  end

  # Test A record creation
  def test_create_a
    expected_name = 'test.example.com'
    expected_address = '10.1.1.1'
    expected_type = 'A'
    record = MmJsonClient::DNSRecord.new(:name => expected_name, :value => expected_address, :type => expected_type)
    response = stub('a_response', :ref => '567')


    @client.stubs(:get_dns_zones).returns(@zones)
    @client.stubs(:match_zone).returns(@zone)
    @client.expects(:add_dns_record).with(has_entry(:dns_record, instance_of(MmJsonClient::DNSRecord))).returns(response)
    assert_equal response,@provider.do_create(expected_name, expected_address, expected_type)
  end
```

**Comment by dmitri-d** [comment]
Ahh, MmJsonClient library doesn't implement equality method on ```MmJsonClient::DNSRecord```.  In this case I would suggest adding a helper method that creates ```MmJsonClient::DNSRecord```, something like:

```
def create_dns_record(name, address, type)
  MmJsonClient::DNSRecord.new(:name => name, :value => address, :type => type)
end
```

then in your test:
```
def test_create_a
    expected_name = 'test.example.com'
    expected_address = '10.1.1.1'
    expected_type = 'A'
    record = MmJsonClient::DNSRecord.new(:name => expected_name, :value => expected_address, :type => expected_type)
    response = stub('a_response', :ref => '567')


    @client.stubs(:get_dns_zones).returns(@zones)
    @client.stubs(:match_zone).returns(@zone)
    @client.expects(:create_dns_record).with(expected_name, expected_address, expected_type).returns(record)
    @client.expects(:add_dns_record).with(has_entry(:dns_record, record)).returns(response)
    assert_equal response,@provider.do_create(expected_name, expected_address, expected_type)
  end
```
